### PR TITLE
Feature/perm group clients (Issue #35)

### PR DESCRIPTION
**Pull request by JimHume**
This PR:
1. Adds a client for group-based administration. 
2. Unit tests for aforementioned group-based client that follow the current pattern in the solution.
3. Extends the IKeycloakProtectedResourceClient functionality to allow for resource deletions and the fetching of all resources.
4. Extends the IKeycloakProtectionClient to get a token for a specific resource and also get the current resource permissions for said resource.
5. Adds extension methods to make consuming this from a combined blazor-server-but-also-a-separate-asp.net-api-project setup simpler (AddKeycloakPolicyHttpClient / AddKeycloakProtectionHttpClient).
6. Adds unit tests for the existing KeycloakProtectedResourceClient class.
7. Allows for pulling the current access token from the http context (AccessTokenPropagationHandler) which is primarily used in blazor server--the access token exists but can be retrieved from the context without the authentication scheme being provided. This happens when attempting to perform actions from a server on behalf of a client (i.e.: blazor server behaves differently than blazor web assembly which I assume is the source of this issue but I could be wrong about its origin).

This also has some minor updates to formatting / styling so that:
1. Classes are in their own files (Owner, ResourceScope, Scope).
2. Namespace / using statements are consistent and reflect class locations.
3. Parameter names and their XML comments are consistent.

TL;DR: this is a rather large PR. It mostly extends rather than altering underlying behaviours that may cause regressions, with the exception of AccessTokenPropagationHandler, but said updates are aggressively wrapped in try/catch statements to ensure the existing behaviour is still exhibited.

### CODE REVIEW

File: src/Keycloak.AuthServices.Common/AuthorizationCollection.cs
Comment (JimHume) [inline src/Keycloak.AuthServices.Common/AuthorizationCollection.cs:1]: This is used to pull a list of claims from the current access token. When issuing a token based on a specific resource, one needs to see what those resources are. The AuthorizationCollection represents a list of ResourcePermission objects that are taken from the Keycloak "permission" definition.

File: src/Keycloak.AuthServices.Common/KeycloakAuthorizationClaimsExtensions.cs
Comment (JimHume) [inline src/Keycloak.AuthServices.Common/KeycloakAuthorizationClaimsExtensions.cs:1]: This is used to pull the ResourcePermissions out of the current access token (claims list, really).

File: src/Keycloak.AuthServices.Sdk/Admin/Constants/KeycloakClientApiConstants.cs
Comment (JimHume) [inline src/Keycloak.AuthServices.Sdk/Admin/Constants/KeycloakClientApiConstants.cs:1]: This doesn't diff too well online.
The TL;DR: is that I added new consts for the groups and policies API's and standardized the path structure within some other consts.

File: src/Keycloak.AuthServices.Sdk/Admin/IKeycloakGroupClient.cs
Comment (JimHume) [inline src/Keycloak.AuthServices.Sdk/Admin/IKeycloakGroupClient.cs:1]: This is the new groups client.

I tried to follow the pattern of other clients in the solution.

File: src/Keycloak.AuthServices.Sdk/Admin/IKeycloakProtectedResourceClient.cs [hunk 1]
```diff
@@ -12,29 +13,40 @@ public interface IKeycloakProtectedResourceClient
...
+
+    /// <summary>
+    /// Searches for resources. NOTE: you must set the <see cref="GetResourcesRequestParameters.Deep"/> property to true
+    /// </summary>
+    /// <param name="realm">Realm name (not ID).</param>
+    /// <param name="getResourcesDeepRequestParameters">Query parameters</param>
+    /// <returns></returns>
+    [Get(KeycloakClientApiConstants.GetResources)]
+    [Headers("Accept: application/json")]
+    Task<List<ResourceResponse>> GetResourcesDeep(string realm, [Query] GetResourcesDeepRequestParameters getResourcesDeepRequestParameters);
```
Comment (JimHume) [inline src/Keycloak.AuthServices.Sdk/Admin/IKeycloakProtectedResourceClient.cs:31]: This is remarkably awkward. The Keycloak API returns different values from the same endpoint based on a query parameter being present. This is why the same endpoint is targeted for both "GET" calls, with the difference being the query parameter object.
You MUST use the new object in order for the 'deep' (thorough response) to be returned. Otherwise the existing behaviour is exhibited.

File: src/Keycloak.AuthServices.Sdk/Admin/IKeycloakUserClient.cs
Comment (JimHume) [inline src/Keycloak.AuthServices.Sdk/Admin/IKeycloakUserClient.cs:1]: This adds the ability to put users in specific groups programmatically via the API.

The existing client was extended to allow for joining, leaving, and listing groups. Related to this: the groups property on the user object was changed from `init` to `set` so that an existing user object could be updated to represent the true state of it from Keycloak (as best as possible).

File: src/Keycloak.AuthServices.Sdk/Admin/Models/Groups/Group.cs [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+public class Group
+{
+    public Dictionary<string, bool>? Access { get; init; }
+    public Dictionary<string, string[]>? Attributes { get; init; }
+    public Dictionary<string, string>? ClientRoles { get; init; }
+    public string? Id { get; init; }
+    public string? Name { get; init; }
+    public string? Path { get; init; }
+    public string[]? RealmRoles { get; init; }
+    public Group[]? SubGroups { get; set; }
```
Comment (JimHume) [inline src/Keycloak.AuthServices.Sdk/Admin/Models/Groups/Group.cs:15]: Subgroups are returned in a separate call.

File: src/Keycloak.AuthServices.Sdk/Admin/Models/Resources/ResourceResponse.cs [hunk 3]
```diff
@@ -20,18 +20,3 @@ public class ResourceResponse
...
-
-public class ResourceScope
-{
-    public string Name { get; set; }
-}
-
-public class Scope
-{
-    public string Name { get; set; }
-}
```
Comment (JimHume) [inline src/Keycloak.AuthServices.Sdk/Admin/Models/Resources/ResourceResponse.cs:37]: These three classes were moved to their own files.

File: src/Keycloak.AuthServices.Sdk/Admin/Requests/Resources/GetResourcesDeepRequestParameters.cs
Comment (JimHume) [inline src/Keycloak.AuthServices.Sdk/Admin/Requests/Resources/GetResourcesDeepRequestParameters.cs:1]: This is the workaround for the Keycloak API returning different values / object structures based on the single query parameter "deep" being set to true.

I wasn't sure how to best resolve the issue from a pattern perspective. The code works without issue but there may be documentation required around this.

File: src/Keycloak.AuthServices.Sdk/AuthZ/IKeycloakPolicyClient.cs
Comment (JimHume) [inline src/Keycloak.AuthServices.Sdk/AuthZ/IKeycloakPolicyClient.cs:1]: This allows one to create resources and then assign policies to them.
This is helpful for enterprise solutions that may create a group hierarchy and want to only allow certain users access to certain resources.

File: src/Keycloak.AuthServices.Sdk/AuthZ/IKeycloakProtectionClient.cs
Comment (JimHume) [inline src/Keycloak.AuthServices.Sdk/AuthZ/IKeycloakProtectionClient.cs:1]: This follows the policies theme: it allows for the issuing of tokens for a particular resource (or set of resources).
It also allows for the querying of all available resources a user has access to. This is useful when attempting to display all resources in a UI e.g.

File: src/Keycloak.AuthServices.Sdk/AuthZ/KeycloakProtectionClient.cs
Comment (JimHume) [inline src/Keycloak.AuthServices.Sdk/AuthZ/KeycloakProtectionClient.cs:1]: The user can attempt to obtain a ticket / token for any given resource. Keycloak will return the expected status codes of forbidden / unauthorized / etc if no such permissions or policies exist and the user does not have access.

I attempted to follow the existing pattern in the file for the new methods.

File: src/Keycloak.AuthServices.Sdk/AuthZ/ServiceCollectionExtensions.cs
Comment (JimHume) [inline src/Keycloak.AuthServices.Sdk/AuthZ/ServiceCollectionExtensions.cs:1]: The new extension methods makes it simpler to consume the library from blazor server.

File: src/Keycloak.AuthServices.Sdk/HttpMiddleware/AccessTokenPropagationHandler.cs [hunk 4]
```diff
@@ -32,8 +32,31 @@ public AccessTokenPropagationHandler(IHttpContextAccessor contextAccessor)
         }
 
         var httpContext = this.contextAccessor.HttpContext;
-        var token = await httpContext
-            .GetTokenAsync(JwtBearerDefaults.AuthenticationScheme, "access_token");
+
```
Comment (JimHume) [inline src/Keycloak.AuthServices.Sdk/HttpMiddleware/AccessTokenPropagationHandler.cs:35]: When attempting to perform actions on behalf of a user in blazor server, the "AuthenticationScheme" does not return the access token. It is available by simply requesting "access_token" without the scheme. 
This attempts to obtain the access token in the secondary fashion if the first attempt does not succeed.
If this second attempt fails then its exception is ignored and the original exception is thrown in order to ensure the existing behaviour is still exhibited and users won't be affected.

File: tests/Keycloak.AuthServices.Sdk.Tests/Admin/KeycloakGroupClientTests.cs
Comment (JimHume) [inline tests/Keycloak.AuthServices.Sdk.Tests/Admin/KeycloakGroupClientTests.cs:1]: I attempted to follow the existing test class structure for the new group client tests.

File: tests/Keycloak.AuthServices.Sdk.Tests/Admin/KeycloakProtectedResourceClientTests.cs
Comment (JimHume) [inline tests/Keycloak.AuthServices.Sdk.Tests/Admin/KeycloakProtectedResourceClientTests.cs:1]: I made (relatively heavy) use of the protected resource client and noticed there were no existing unit tests. I added a new set of tests that follow the pattern of other unit tests in the solution.

File: tests/Keycloak.AuthServices.Sdk.Tests/AuthZ/KeycloakPolicyClientTests.cs
Comment (JimHume) [inline tests/Keycloak.AuthServices.Sdk.Tests/AuthZ/KeycloakPolicyClientTests.cs:1]: Ditto this for unit tests / newly-added / follow existing pattern.

**Comment by JimHume** [comment]
TL;DR: (again): I added comments to some files to hopefully clarify some of the intentions of the code changes. 

I believe the PR is now ready for review.

**Comment by NikiforovAll** [comment]
Thanks, @JimHume , I will take a look at this. 🙏

**Comment by NikiforovAll** [comment]
@JimHume I plan to review this. Would you like to have a call so we can work on it together?

**Comment by JimHume** [comment]
@NikiforovAll Indeed. I will try to reach you via your email link on your github profile page.

**Comment by NikiforovAll** [comment]
I plan to move to automatic client generation based on OpenAPI Spec. So this ticket is no longer needed.

**Comment by NikiforovAll** [comment]
Since the quality of auto-generated code is low, I decided to keep the previous approach. Although, we might want to migrate from Refit, which should not be hard to to do.

**Comment by NikiforovAll** [comment]
The original contribution was partially included in version 2.0.0. I appreciate your effort and thank you for your contributions! If you believe that some features from the previous version are missing in version 2.0.0, I encourage you to consider re-implementing those features in the new version. I apologize for any inconvenience this may cause.
